CesiumGltf/include/CesiumGltf/PropertyTypeTraits.h

```diff
@@ -0,0 +1,139 @@
+
+namespace CesiumGltf {
+/**
+ * @brief Check if a C++ type can be represented as a numeric property type
+ */
+template <typename... T> struct IsMetadataNumeric;
+template <typename T> struct IsMetadataNumeric<T> : std::false_type {};
+template <> struct IsMetadataNumeric<uint8_t> : std::true_type {};
+template <> struct IsMetadataNumeric<int8_t> : std::true_type {};
+template <> struct IsMetadataNumeric<uint16_t> : std::true_type {};
+template <> struct IsMetadataNumeric<int16_t> : std::true_type {};
+template <> struct IsMetadataNumeric<uint32_t> : std::true_type {};
+template <> struct IsMetadataNumeric<int32_t> : std::true_type {};
+template <> struct IsMetadataNumeric<uint64_t> : std::true_type {};
+template <> struct IsMetadataNumeric<int64_t> : std::true_type {};
+template <> struct IsMetadataNumeric<float> : std::true_type {};
+template <> struct IsMetadataNumeric<double> : std::true_type {};
+
+/**
+ * @brief Check if a C++ type can be represented as a boolean property type
+ */
+template <typename... T> struct IsMetadataBoolean;
+template <typename T> struct IsMetadataBoolean<T> : std::false_type {};
+template <> struct IsMetadataBoolean<bool> : std::true_type {};
+
+/**
+ * @brief Check if a C++ type can be represented as a string property type
+ */
+template <typename... T> struct IsMetadataString;
+template <typename T> struct IsMetadataString<T> : std::false_type {};
+template <> struct IsMetadataString<std::string_view> : std::true_type {};
+
+/**
+ * @brief Check if a C++ type can be represented as an array of number property
+ * type
+ */
+template <typename... T> struct IsMetadataNumericArray;
+template <typename T> struct IsMetadataNumericArray<T> : std::false_type {};
+template <typename T> struct IsMetadataNumericArray<MetadataArrayView<T>> {
+  static constexpr bool value = IsMetadataNumeric<T>::value;
+};
+
+/**
+ * @brief Check if a C++ type can be represented as an array of boolean property
+ * type
+ */
+template <typename... T> struct IsMetadataBooleanArray;
+template <typename T> struct IsMetadataBooleanArray<T> : std::false_type {};
+template <>
+struct IsMetadataBooleanArray<MetadataArrayView<bool>> : std::true_type {};
+
+/**
+ * @brief Check if a C++ type can be represented as an array of string property
+ * type
+ */
+template <typename... T> struct IsMetadataStringArray;
+template <typename T> struct IsMetadataStringArray<T> : std::false_type {};
+template <>
+struct IsMetadataStringArray<MetadataArrayView<std::string_view>>
+    : std::true_type {};
+
+/**
+ * @brief Retrieve the component type of a metadata array
+ */
+template <typename T> struct MetadataArrayType;
+template <typename T>
+struct MetadataArrayType<CesiumGltf::MetadataArrayView<T>> {
+  using type = T;
+};
+
+/**
+ * @brief Convert a C++ type to PropertyType
+ */
+template <typename T> struct TypeToPropertyType;
+
+template <> struct TypeToPropertyType<uint8_t> {
+  static constexpr uint32_t value = static_cast<uint32_t>(PropertyType::Uint8);
```

Comment:
Why cast to `uint32_t`, rather than just using `static constexpr PropertyType value = PropertyType::Uint8;` ? It looks like most uses of these templates are comparing against the enum type, and this formulation requires a cast there, too.

Reply:
Initially, the main use of uint32_t is for metadata array mostly. For example, it maybe helpful to know if the type is Array and what is the component of the array as well. I use it in the initial implementations of PropertyView, but maybe not needed right now. Do you think if the use case above is valid or should I convert uint32_t to PropertyType?

Reply:
I'm not sure what you mean here. How does making `value` a uint32_t instead of a `PropertyType` help us "know if the type is Array and what is the component of the array as well"?

If there are places where you need the PropertyType value as a uint32_t, you can add the cast there. It seems like it would be better for this mapping to return the strongly-typed value (PropertyType). But I could very well be missing something.

Reply:
Oh I mean it uses bitmasking to know the type of the array. For example, `MetadataArray<int>` will be converted to uint32_t `uint32_t(PropertyType::Array) | uint32_t(PropertyType::Int)`, so client only checks if the mask contains `Array` and `Int` bits to see the array type. But that's not really used anywhere though and probably not useful, so I think I may use PropertyType directly for simplicity